staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go

```diff
@@ -183,73 +183,141 @@ func TestListenerResyncPeriods(t *testing.T) {
+	if e, a := zeroDuration(1*time.Minute, !doResyncs), informer.resyncCheckPeriod; e != a {
 		t.Errorf("expected %d, got %d", e, a)
 	}
-	if e, a := time.Duration(0), informer.processor.listeners[0].resyncPeriod; e != a {
+	if e, a := time.Duration(0), informer.processor.listeners[0].requestedResyncPeriod; e != a {
 		t.Errorf("expected %d, got %d", e, a)
 	}
-	if e, a := 1*time.Minute, informer.processor.listeners[1].resyncPeriod; e != a {
+	if e, a := 1*time.Minute, informer.processor.listeners[1].requestedResyncPeriod; e != a {
 		t.Errorf("expected %d, got %d", e, a)
 	}
+	checkResyncPeriods(t, informer.processor, !doResyncs)
 
 	// listener 3, resync every 55 seconds
-	listener3 := newTestListener("listener3", 55*time.Second)
+	listener3 := newTestListener("listener3", 55*time.Second, "pod1")
 	informer.AddEventHandlerWithResyncPeriod(listener3, listener3.resyncPeriod)
-	if e, a := 55*time.Second, informer.resyncCheckPeriod; e != a {
+	if e, a := zeroDuration(55*time.Second, !doResyncs), informer.resyncCheckPeriod; e != a {
 		t.Errorf("expected %d, got %d", e, a)
 	}
-	if e, a := time.Duration(0), informer.processor.listeners[0].resyncPeriod; e != a {
+	if e, a := time.Duration(0), informer.processor.listeners[0].requestedResyncPeriod; e != a {
 		t.Errorf("expected %d, got %d", e, a)
 	}
-	if e, a := 1*time.Minute, informer.processor.listeners[1].resyncPeriod; e != a {
+	if e, a := 1*time.Minute, informer.processor.listeners[1].requestedResyncPeriod; e != a {
 		t.Errorf("expected %d, got %d", e, a)
 	}
-	if e, a := 55*time.Second, informer.processor.listeners[2].resyncPeriod; e != a {
+	if e, a := 55*time.Second, informer.processor.listeners[2].requestedResyncPeriod; e != a {
 		t.Errorf("expected %d, got %d", e, a)
 	}
+	checkResyncPeriods(t, informer.processor, !doResyncs)
 
 	// listener 4, resync every 5 seconds
-	listener4 := newTestListener("listener4", 5*time.Second)
+	listener4 := newTestListener("listener4", 5*time.Second, "pod1")
 	informer.AddEventHandlerWithResyncPeriod(listener4, listener4.resyncPeriod)
-	if e, a := 5*time.Second, informer.resyncCheckPeriod; e != a {
+	if e, a := zeroDuration(5*time.Second, !doResyncs), informer.resyncCheckPeriod; e != a {
+		t.Errorf("expected %d, got %d", e, a)
+	}
+	if e, a := time.Duration(0), informer.processor.listeners[0].requestedResyncPeriod; e != a {
+		t.Errorf("expected %d, got %d", e, a)
+	}
+	if e, a := 1*time.Minute, informer.processor.listeners[1].requestedResyncPeriod; e != a {
+		t.Errorf("expected %d, got %d", e, a)
+	}
+	if e, a := 55*time.Second, informer.processor.listeners[2].requestedResyncPeriod; e != a {
+		t.Errorf("expected %d, got %d", e, a)
+	}
+	if e, a := 5*time.Second, informer.processor.listeners[3].requestedResyncPeriod; e != a {
+		t.Errorf("expected %d, got %d", e, a)
+	}
+	checkResyncPeriods(t, informer.processor, !doResyncs)
+
+	// Now start the informer so we can test that the resyncCheckPeriod shrinks no further
```

Comment:
it seems like we could have easier ways to unit test this.

Reply:
Could you elaborate on what you mean by "easier" and why it would be better?

Reply:
I'm not entirely sure I understand what this is testing, but I didn't write the shared informer. Does each listener not get their exact resync period or something?

Reply:
Yes, that's part of the point here.  Each listener added before the informer starts gets their exact resync period --- if the informer does resyncs at all; each listener added after the informer starts is constrained by a configuration choice frozen at start time.

Overall this is testing the resync periods delivered to each client, as well as some details of the supporting data structure.  The point of the latter is to confirm the redundancy in the data structure (#86862), which I will then submit a PR to remove.